File: Swift/TicTacToeSolver/Negamax.swift
//
//  Negamax.swift
//  TicTacToeSolver
//
//  Created by Daniel Seitz on 8/8/16.
//  Copyright © 2016 Daniel Seitz. All rights reserved.
//

import Foundation

func negamax(onmove: Int, board: Board) -> Int {
  let n = board.count
  var value = gamevalue(onmove, board: board)
  if value != -2 {
    return value
  }
  value = -1
  for row in 0..<n {
    for col in 0..<n {
      if board[row][col] == 0 {
        board[row][col] = onmove
        let v0 = -negamax(-onmove, board: board)
        if v0 > value {
          value = v0
        }
        board[row][col] = 0
      }
    }
  }
  return value
}

func negamax_symmetry(onmove: Int, board: Board, zobrist: Zobrist) -> Int {
  let n = board.count
  if let calculated = zobrist.checkSymmetry(board) {
    return calculated
  }
  var value = gamevalue(onmove, board: board)
  if value != -2 {
    return value
  }
  for row in 0..<n {
    for col in 0..<n {
      if board[row][col] == 0 {
        board[row][col] = onmove
        let v0 = -negamax_symmetry(-onmove, board: board, zobrist: zobrist)
        if v0 > value {
          value = v0
        }
        board[row][col] = 0
      }
    }
  }
  zobrist.store(value: value, for: board)
  return value
}

func negamax_alpha_beta(onmove: Int, board: Board, alpha: Int, beta: Int) -> Int {
  let n = board.count
  var value = gamevalue(onmove, board: board)
  if value != -2 {
    return value
  }
  var alpha = alpha
  for row in 0..<n {
    for col in 0..<n {
      if board[row][col] == 0 {
        board[row][col] = onmove
        let v0 = -negamax_alpha_beta(-onmove, board: board, alpha: -beta, beta: -alpha)
        if v0 > value {
          value = v0
        }
        if v0 > alpha {
          alpha = v0
        }
        board[row][col] = 0
        if alpha >= beta {
          return alpha
        }
      }
    }
  }
  return value
}

func negamax_both(onmove: Int, board: Board, alpha: Int, beta: Int, zobrist: Zobrist) -> Int {
  let n = board.count
  if let calculated = zobrist.checkSymmetry(board) {
    return calculated
  }
  var value = gamevalue(onmove, board: board)
  if value != -2 {
    zobrist.store(value: value, for: board)
    return value
  }
  var alpha = alpha
  for row in 0..<n {
    for col in 0..<n {
      if board[row][col] == 0 {
        board[row][col] = onmove
        let v0 = -negamax_both(-onmove, board: board, alpha: -beta, beta: -alpha, zobrist: zobrist)
        if v0 > value {
          value = v0
        }
        if v0 > alpha {
          alpha = v0
        }
        if alpha >= beta {
          zobrist.store(value: alpha, for: board)
          board[row][col] = 0
          return alpha
        }
        board[row][col] = 0
      }
    }
  }
  zobrist.store(value: value, for: board)
  return value
}

File: Rust/TicTacToeSolver/src/negamax.rs
use board::Board;
use zobrist::Zobrist;

pub fn negamax(onmove: i32, board: &mut Board) -> i32 {
  let n = board.n as usize;
  let mut value = board.gamevalue(onmove);
  if value == -2 {
    value = -1;
    for row in 0..n {
      for col in 0..n {
        board.try_move(onmove, row, col, |b| {
            let v0 = -negamax(-onmove, b);
            if v0 > value {
              value = v0;
            }
          });
      }
    }
  }
  value
}

pub fn negamax_symmetry(onmove: i32, board: &mut Board, zobrist: &mut Zobrist) -> i32 {
  let n = board.n as usize;
  if let Some(value) = zobrist.check_symmetry(board) {
    return value;
  }
  let mut value = board.gamevalue(onmove);
  if value == -2 {
    value = -1;
    for row in 0..n {
      for col in 0..n {
        board.try_move(onmove, row, col, |b| {
          let v0 = -negamax_symmetry(-onmove, b, zobrist);
          if v0 > value {
            value = v0;
          }
        });
      }
    }
  }
  zobrist.store(value, board);
  value
}

pub fn negamax_alpha_beta(onmove: i32, board: &mut Board, mut alpha: i32, beta: i32) -> i32 {
  let n = board.n as usize;
  let mut value = board.gamevalue(onmove);
  if value == -2 {
    for row in 0..n {
      for col in 0..n {
        board.try_move(onmove, row, col, |b| {
          let v0 = -negamax_alpha_beta(-onmove, b, -beta, -alpha);
          if v0 > value {
            value = v0;
          }
          if v0 > alpha {
            alpha = v0;
          }
        });
        if alpha >= beta {
          return alpha;
        }
      }
    }
  }
  value
}

pub fn negamax_both(onmove: i32, board: &mut Board, mut alpha: i32, beta: i32, zobrist: &mut Zobrist) -> i32 {
  let n = board.n as usize;
  if let Some(value) = zobrist.check_symmetry(board) {
    return value;
  }
  let mut value = board.gamevalue(onmove);
  if value == -2 {
    for row in 0..n {
      for col in 0..n {
        board.try_move(onmove, row, col, |b| {
          let v0 = -negamax_both(-onmove, b, -beta, -alpha, zobrist);
          if v0 > value {
            value = v0;
          }
          if v0 > alpha {
            alpha = v0;
          }
          if alpha >= beta {
            zobrist.store(alpha, b);
          }
        });
        if alpha >= beta {
          return alpha;
        }
      }
    }
  }
  zobrist.store(value, board);
  value
}
